README.md
# dsRedactorExtends plugin for Craft CMS

This is intended to add Redactor Plugins.

## Installation

To install dsRedactorExtends, follow these steps:

1. Download & unzip the file and place the `dsredactorextends` directory into your `craft/plugins` directory
2.  -OR- do a `git clone https://github.com/dreamseeker/dsredactorextends.git` directly into your `craft/plugins` folder.  You can then update it with `git pull`
3.  -OR- install with Composer via `composer require dreamseeker/dsredactorextends`
4. Install plugin in the Craft Control Panel under Settings > Plugins
5. The plugin folder should be named `dsredactorextends` for Craft to see it.  GitHub recently started appending `-master` (the branch name) to the name of the folder for zip file downloads.

dsRedactorExtends works on Craft 2.6.x. and Redactor II.

## dsRedactorExtends Overview

dsRedactorExtends adds following Redactor Plugins to RichText Editor.

* Inline Style  
  https://imperavi.com/redactor/plugins/inlinestyle/
* Font Color  
  https://imperavi.com/redactor/plugins/font-color/
* Font Size  
  https://imperavi.com/redactor/plugins/font-size/
* Underline   
  https://imperavi.com/redactor/examples/button-underline/
* Clips   
  https://imperavi.com/redactor/plugins/clips/

## Using dsRedactorExtends

Please create .json files into `craft/config/redactor` for Redactor config.

**JSON Sample**

```
{
  plugins: ['inlinestyle', 'fontcolor', 'fontsize', 'underline', 'clips'],
  inlinestyleTags: {
    "style1": {
      title: "InlineStyle01",
      args: ['mark', 'class', 'styleA', 'toggle']
    },
    "style2": {
      title: "InlineStyle02",
      args: ['code', 'class', 'styleB', 'toggle']
    },
  },
  fontcolorLists: [
    '#ffffff', '#000000', '#eeece1', '#1f497d', '#4f81bd', '#c0504d', '#9bbb59', '#8064a2', '#4bacc6', '#f79646', '#ffff00',
    '#f2f2f2', '#7f7f7f', '#ddd9c3', '#c6d9f0', '#dbe5f1', '#f2dcdb', '#ebf1dd', '#e5e0ec', '#dbeef3', '#fdeada', '#fff2ca',
    '#d8d8d8', '#595959', '#c4bd97', '#8db3e2', '#b8cce4', '#e5b9b7', '#d7e3bc', '#ccc1d9', '#b7dde8', '#fbd5b5', '#ffe694'
  ],
  bgcolorLists: [
    '#ffffff', '#000000', '#eeece1', '#1f497d', '#4f81bd', '#c0504d', '#9bbb59', '#8064a2', '#4bacc6', '#f79646', '#ffff00'
  ],
  fontSize: [10, 12, 14, 16, 18, 20],
  clipsItems: [
      ['snippet1', '<h1>Heading Text</h1>'],
      ['snippet2', '<p>Body Text.</p>'],
  }
}
```

When you apply custom CSS to Redacgtor editing area, please appoint a file path by Plugin setting.

**CSS Sample**

```
.redactor-layer h1 {
  color: #c00;
}
.redactor-layer p {
  color: #333;
}
```

> [hint]  Redactor editing area is wrapped by `redactor-layer` class.

## dsRedactorExtends Changelog

## 1.0.2 -- 2017.07.05

* [add] one plugin
  * Clips
* [add] Plugin Setting to apply custom CSS

### 1.0.1 -- 2017.06.24

* [add] two plugins
  * Font Size
  * Underline

### 1.0.0 -- 2017.06.22

* Release it

Brought to you by [dreamseeker](https://github.com/dreamseeker)


CHANGELOG.md
# dsRedactorExtends Changelog

## 1.0.2 -- 2017.07.05

* [add] one plugin
  * Clips
* [add] Plugin Setting to apply custom CSS

## 1.0.1 -- 2017.06.24

* [add] two plugins
  * Font Size
  * Underline
  
## 1.0.0 -- 2017.06.22

* Initial release

Brought to you by [dreamseeker](https://github.com/dreamseeker)


resources/plugins/clips.js
// Clips
// Adapted from offical Redactor site
// https://imperavi.com/redactor/plugins/clips/

if (!RedactorPlugins) var RedactorPlugins = {};

RedactorPlugins.clips = function () {
  return {
    langs: {
      en: {
        "insertclips": "Insert Clips",
        "textcolor": "Text",
        "highlight": "Highlight",
        "resetcolor": "none"
      },
      ja: {
        "insertclips": "定型文の挿入",
        "textcolor": "文字色",
        "highlight": "背景色",
        "resetcolor": "リセット"
      }
    },
    init: function () {
      var items = [
        ['Lorem ipsum...', 'Lorem ipsum dolor sit amet, consectetur adipisicing elit, sed do eiusmod tempor incididunt ut labore et dolore magna aliqua. Ut enim ad minim veniam, quis nostrud exercitation ullamco laboris nisi ut aliquip ex ea commodo consequat. Duis aute irure dolor in reprehenderit in voluptate velit esse cillum dolore eu fugiat nulla pariatur. Excepteur sint occaecat cupidatat non proident, sunt in culpa qui officia deserunt mollit anim id est laborum.'],
        ['Red label', '<b class="label-red">Label</b>']
      ];

      // if define custom colors, override them.
      if (this.opts.clipsItems) {
        items = this.opts.clipsItems;
      }

      this.clips.template = $('<ul id="redactor-modal-list">');

      for (var i = 0; i < items.length; i++) {
        var li = $('<li>');
        var a = $('<a href="#" class="redactor-clips-link">').html(items[i][0]);
        var div = $('<div class="redactor-clips">').hide().html(items[i][1]);

        li.append(a);
        li.append(div);
        this.clips.template.append(li);
      }

      this.modal.addTemplate('clips', '<div class="modal-section">' + this.utils.getOuterHtml(this.clips.template) + '</div>');

      var button = this.button.add('clips', 'Clips');
      this.button.setIcon(button, '<i class="re-icon-clips"></i>');
      this.button.addCallback(button, this.clips.show);

    },
    show: function () {
      this.modal.load('clips', this.lang.get('insertclips'), 500);

      $('#redactor-modal-list').find('.redactor-clips-link').each($.proxy(this.clips.load, this));

      this.modal.show();
    },
    load: function (i, s) {
      $(s).on('click', $.proxy(function (e) {
        e.preventDefault();
        this.clips.insert($(s).next().html());

      }, this));
    },
    insert: function (html) {
      this.buffer.set();
      this.air.collapsedEnd();
      this.insert.html(html);
      this.modal.close();
    }
  };
};



DsRedactorExtendsPlugin.php
<?php
/**
 * dsRedactorExtends plugin for Craft CMS
 *
 * This is intended to add Redactor Plugins.
 *
 * @author    dreamseeker
 * @copyright Copyright (c) 2017 dreamseeker
 * @link      https://github.com/dreamseeker
 * @package   DsRedactorExtends
 * @since     1.0.0
 */

namespace Craft;

class DsRedactorExtendsPlugin extends BasePlugin
{
    /**
     * @return mixed
     */
    public function init()
    {
        parent::init();

      if (craft()->request->isCpRequest()) {
        craft()->templates->includeJsResource('dsredactorextends/plugins/inlinestyle.js');
        craft()->templates->includeJsResource('dsredactorextends/plugins/fontcolor.js');
        craft()->templates->includeJsResource('dsredactorextends/plugins/fontsize.js');
        craft()->templates->includeJsResource('dsredactorextends/plugins/underline.js');
        craft()->templates->includeJsResource('dsredactorextends/plugins/clips.js');
        // add custom CSS
        $this->_renderCss();
      }
    }

    /**
     * @return mixed
     */
    public function getName()
    {
         return Craft::t('dsRedactorExtends');
    }

    /**
     * @return mixed
     */
    public function getDescription()
    {
        return Craft::t('This is intended to add Redactor Plugins.');
    }

    /**
     * @return string
     */
    public function getDocumentationUrl()
    {
        return 'https://github.com/dreamseeker/craft-dsRedactorExtends/blob/master/README.md';
    }

    /**
     * @return string
     */
    public function getReleaseFeedUrl()
    {
        return 'https://raw.githubusercontent.com/dreamseeker/craft-dsRedactorExtends/master/releases.json';
    }

    /**
     * @return string
     */
    public function getVersion()
    {
        return '1.0.2';
    }

    /**
     * @return string
     */
    public function getSchemaVersion()
    {
        return '1.0.1';
    }

    /**
     * @return string
     */
    public function getDeveloper()
    {
        return 'dreamseeker';
    }

    /**
     * @return string
     */
    public function getDeveloperUrl()
    {
        return 'https://github.com/dreamseeker';
    }

    /**
     * @return bool
     */
    public function hasCpSection()
    {
        return false;
    }

    /**
     */
    public function onBeforeInstall()
    {
    }

    /**
     */
    public function onAfterInstall()
    {
    }

    /**
     */
    public function onBeforeUninstall()
    {
    }

    /**
     */
    public function onAfterUninstall()
    {
    }

    /**
     * plugin settings
     */
    protected function defineSettings()
    {
        return array(
            'cssFilePath' => array(AttributeType::String),
        );
    }

    /**
     * Setting Page HTML
     */
    public function getSettingsHtml()
    {
        return craft()->templates->render('dsredactorextends/_settings', array(
            'settings' => $this->getSettings(),
        ));
    }

    /**
     * render Custom CSS
     */
    private function _renderCss()
    {
        $settings = $this->getSettings();
        if (trim($settings->cssFilePath)) {
            $filepath = craft()->config->parseEnvironmentString($settings->cssFilePath);
            if ($hash = @sha1_file($filepath)) {
                craft()->templates->includeCssFile($filepath.'?e='.$hash);
            } else {
                craft()->templates->includeCssFile($filepath);
            }
        }
    }
}


resources/plugins/fontsize.js
// Font Size
// Adapted from offical Redactor site
// https://imperavi.com/redactor/plugins/font-size/

if (!RedactorPlugins) var RedactorPlugins = {};

RedactorPlugins.fontsize = function () {
  return {
    langs: {
      en: {
        "Size": "Size",
      },
      ja: {
        "Size": "文字サイズ",
      }
    },
    init: function () {
      var fontSize = [10, 11, 12, 14, 16, 18, 20, 24, 28, 30];
      var that = this;
      var dropdown = {};
      
      // if define custom fontSize, override them.
      if (this.opts.fontSize) {
        fontSize = this.opts.fontSize;
      }

      $.each(fontSize, function (i, s) {
        dropdown['s' + i] = {
          title: s + 'px', func: function () {
            that.fontsize.set(s);
          }
        };
      });

      dropdown.remove = {title: 'Remove Font Size', func: that.fontsize.reset};

      var button = this.button.addBefore('format', 'fontsize', this.lang.get('Size'));
      this.button.setIcon(button, '<i class="re-icon-fontsize"></i>');
      this.button.addDropdown(button, dropdown);
    },
    set: function (size) {
      this.inline.format('span', 'style', 'font-size: ' + size + 'px;');
    },
    reset: function () {
      this.inline.removeStyleRule('font-size');
    }
  };
};


resources/plugins/inlinestyle.js
// Inline Style
// Adapted from offical Redactor site
// https://imperavi.com/redactor/plugins/inlinestyle/

if (!RedactorPlugins) var RedactorPlugins = {};

RedactorPlugins.inlinestyle = function () {
  return {
    langs: {
      en: {
        "style": "Style"
      },
      ja: {
        "style": "インラインスタイル"
      }
    },
    init: function () {
      var tags = {
        "marked": {
          title: "Marked",
          args: ['mark']
        },
        "code": {
          title: "Code",
          args: ['code']
        },
        "sample": {
          title: "Sample",
          args: ['samp']
        },
        "variable": {
          title: "Variable",
          args: ['var']
        },
        "shortcut": {
          title: "Shortcut",
          args: ['kbd']
        },
        "cite": {
          title: "Cite",
          args: ['cite']
        },
        "sup": {
          title: "Superscript",
          args: ['sup']
        },
        "sub": {
          title: "Subscript",
          args: ['sub']
        }
      };

      // if define custom tags, override them.
      if (this.opts.inlinestyleTags) {
        tags = this.opts.inlinestyleTags;
      }

      var that = this;
      var dropdown = {};

      $.each(tags, function (i, s) {
        dropdown[i] = {title: s.title, func: 'inline.format', args: s.args};
      });

      var button = this.button.addAfter('format', 'inline', this.lang.get('style'));
      this.button.setIcon(button, '<i class="re-icon-inline"></i>');
      this.button.addDropdown(button, dropdown);
    }
  };
}


resources/plugins/fontcolor.js
// Font Color
// Adapted from offical Redactor site
// https://imperavi.com/redactor/plugins/font-color/

if (!RedactorPlugins) var RedactorPlugins = {};

RedactorPlugins.fontcolor = function () {
  return {
    langs: {
      en: {
        "textcolorbutton": "Text Color",
        "textcolor": "Text",
        "highlight": "Highlight",
        "resetcolor": "none"
      },
      ja: {
        "textcolorbutton": "文字 / 背景色",
        "textcolor": "文字色",
        "highlight": "背景色",
        "resetcolor": "リセット"
      }
    },
    init: function () {
      var colors,
          bgcolors = [
        '#ffffff', '#000000', '#eeece1', '#1f497d', '#4f81bd', '#c0504d', '#9bbb59', '#8064a2', '#4bacc6', '#f79646', '#ffff00',
        '#f2f2f2', '#7f7f7f', '#ddd9c3', '#c6d9f0', '#dbe5f1', '#f2dcdb', '#ebf1dd', '#e5e0ec', '#dbeef3', '#fdeada', '#fff2ca',
        '#d8d8d8', '#595959', '#c4bd97', '#8db3e2', '#b8cce4', '#e5b9b7', '#d7e3bc', '#ccc1d9', '#b7dde8', '#fbd5b5', '#ffe694',
        '#bfbfbf', '#3f3f3f', '#938953', '#548dd4', '#95b3d7', '#d99694', '#c3d69b', '#b2a2c7', '#b7dde8', '#fac08f', '#f2c314',
        '#a5a5a5', '#262626', '#494429', '#17365d', '#366092', '#953734', '#76923c', '#5f497a', '#92cddc', '#e36c09', '#c09100',
        '#7f7f7f', '#0c0c0c', '#1d1b10', '#0f243e', '#244061', '#632423', '#4f6128', '#3f3151', '#31859b', '#974806', '#7f6000'
      ];
      
      // if define custom colors, override them.
      if (this.opts.fontcolorLists) {
        colors = this.opts.fontcolorLists;
      }

      // if define custom bgcolors, override them.
      if (this.opts.bgcolorLists) {
        bgcolors = this.opts.bgcolorLists;
      }

      var $button = this.button.addBefore('format', 'fontcolor', this.lang.get('textcolorbutton'));
      this.button.setIcon($button, '<i class="re-icon-fontcolor"></i>');

      var $dropdown = this.button.addDropdown($button);
      $dropdown.attr('rel', 'fontcolor');
      $dropdown.width(242);

      var $selector = $('<div style="overflow: hidden; text-align: center;">');
      var $selectorText = $('<span rel="text" class="re-dropdown-box-selector-font" style="background: #eee; float: left; padding: 8px 0; cursor: pointer; font-size: 12px; width: 50%;">' + this.lang.get('textcolor') + '</span>');
      var $selectorBack = $('<span rel="back" class="re-dropdown-box-selector-font" style="float: left; padding: 8px 0; cursor: pointer; font-size: 12px; width: 50%;">' + this.lang.get('highlight') + '</span>');

      $selector.append($selectorText);
      $selector.append($selectorBack);

      $dropdown.append($selector);

      this.fontcolor.buildPicker($dropdown, 'textcolor', colors);
      this.fontcolor.buildPicker($dropdown, 'backcolor', bgcolors);

      $selectorText.on('mousedown', function (e) {
        e.preventDefault();

        $dropdown.find('.re-dropdown-box-selector-font').css('background', 'none');
        $dropdown.find('.re-dropdown-box-backcolor').hide();
        $dropdown.find('.re-dropdown-box-textcolor').show();

        $(this).css('background', '#eee');
      });

      $selectorBack.on('mousedown', function (e) {
        e.preventDefault();

        $dropdown.find('.re-dropdown-box-selector-font').css('background', 'none');
        $dropdown.find('.re-dropdown-box-textcolor').hide();
        $dropdown.find('.re-dropdown-box-backcolor').show();

        $(this).css('background', '#eee');
      });

    },
    buildPicker: function ($dropdown, name, colors) {
      var $box = $('<div class="re-dropdown-box-' + name + '">');
      var rule = (name == 'backcolor') ? 'background-color' : 'color';
      var len = colors.length;
      var self = this;
      var func = function (e) {
        e.preventDefault();
        self.fontcolor.set($(this).data('rule'), $(this).attr('rel'));
      };

      for (var z = 0; z < len; z++) {
        var color = colors[z];

        var $swatch = $('<a rel="' + color + '" data-rule="' + rule + '" href="#" style="float: left; box-sizing: border-box; font-size: 0; border: 2px solid #fff; padding: 0; margin: 0; width: 22px; height: 22px;"></a>');
        $swatch.css('background-color', color);
        $swatch.on('mousedown', func);

        $box.append($swatch);
      }

      var $elNone = $('<a href="#" style="display: block; clear: both; padding: 8px 5px; box-sizing: border-box; font-size: 12px; line-height: 1;"></a>').html(this.lang.get('resetcolor'));
      $elNone.on('mousedown', $.proxy(function (e) {
        e.preventDefault();
        this.fontcolor.remove(rule);

      }, this));

      $box.append($elNone);
      $dropdown.append($box);

      if (name == 'backcolor') {
        $box.hide();
      }
    },
    set: function (rule, type) {
      this.inline.format('span', 'style', rule + ': ' + type + ';');
      this.dropdown.hide();
    },
    remove: function (rule) {
      this.inline.removeStyleRule(rule);
      this.dropdown.hide();
    }
  };
};
